lib/ui/painting/display_list.cc

```diff
@@ -0,0 +1,293 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/lib/ui/painting/display_list.h"
+
+#include "flutter/flow/display_list_interpreter.h"
+#include "flutter/fml/logging.h"
+#include "flutter/fml/make_copyable.h"
+#include "flutter/lib/ui/painting/canvas.h"
+#include "flutter/lib/ui/painting/image_filter.h"
+#include "flutter/lib/ui/painting/path.h"
+#include "flutter/lib/ui/painting/shader.h"
+#include "flutter/lib/ui/ui_dart_state.h"
+#include "third_party/skia/include/core/SkColorFilter.h"
+#include "third_party/skia/include/core/SkImageFilter.h"
+#include "third_party/skia/include/core/SkMaskFilter.h"
+#include "third_party/skia/include/core/SkShader.h"
+#include "third_party/tonic/converter/dart_converter.h"
+#include "third_party/tonic/dart_binding_macros.h"
+#include "third_party/tonic/dart_library_natives.h"
+
+namespace flutter {
+
+static void DisplayList_constructor(Dart_NativeArguments args) {
+  UIDartState::ThrowIfUIOperationsProhibited();
+  DartCallConstructor(&DisplayList::Create, args);
+}
+
+IMPLEMENT_WRAPPERTYPEINFO(ui, DisplayList);
+
+#define FOR_EACH_BINDING(V) V(DisplayList, toImage)
+
+FOR_EACH_BINDING(DART_NATIVE_CALLBACK)
+
+void DisplayList::RegisterNatives(tonic::DartLibraryNatives* natives) {
+  natives->Register(
+      {{"DisplayList_constructor", DisplayList_constructor, 6, true},
+       FOR_EACH_BINDING(DART_REGISTER_NATIVE)});
+}
+
+fml::RefPtr<DisplayList> DisplayList::Create(tonic::Uint8List& ops,
+                                             int numOps,
+                                             tonic::DartByteData& data,
+                                             int numData,
+                                             Dart_Handle objList) {
+  if (numOps < 0 || numOps > ops.num_elements() || numData < 0 ||
+      (data.length_in_bytes() % sizeof(float)) != 0 ||
+      numData > (int)(data.length_in_bytes() / sizeof(float))) {
+    Dart_ThrowException(
+        tonic::ToDart("DisplayList constructor called with bad list lengths."));
+    return nullptr;
+  }
+  if (Dart_IsNull(objList) || !Dart_IsList(objList)) {
+    Dart_ThrowException(
+        tonic::ToDart("DisplayList constructor called with bad object array."));
+    return nullptr;
+  }
+
+  const uint8_t* ops_ptr = ops.data();
+  std::shared_ptr<std::vector<uint8_t>> ops_vector =
+      std::make_shared<std::vector<uint8_t>>(ops_ptr, ops_ptr + numOps);
+
+  const float* data_ptr = static_cast<const float*>(data.data());
+  std::shared_ptr<std::vector<float>> data_vector =
+      std::make_shared<std::vector<float>>(data_ptr, data_ptr + numData);
+
+  intptr_t numObjects = 0;
+  Dart_ListLength(objList, &numObjects);
+  Dart_Handle objects[numObjects];
+  if (Dart_IsError(Dart_ListGetRange(objList, 0, numObjects, objects))) {
+    return nullptr;
+  }
```

Comment:
Consider logging the error, at least `DLOG`.

Reply:
I beefed up the logging and error handling on this section a lot. I wanted to throw a Dart exception, but that doesn't seem to propagate from this code so I gave up and LOG/return empty DL.